Passing translations to LLM via JSON

gambon2010 opened the pull request:
Add logic to translate using structured outputs

taahamahdi commented on src/generate_json/generate_json.ts:
```diff
@@ -0,0 +1,927 @@
+    return batchVerifyItemArray;
+}
+
+function removeLine(): void {
```
Is this overwriting the current output so the log file doesn't get too long?

gambon2010 replied:
Yes

taahamahdi commented on src/generate_json/generate_json.ts:
```diff
@@ -0,0 +1,927 @@
+}
+
+function removeLastPrintCompletion(firstPrint: boolean): void {
+    if (!firstPrint) {
```
Why six calls in particular? Is there a cleaner way to do this?

gambon2010 replied:
I'm printing 6 lines. There probably is a cleaner way, but I didn't look too much into this yet. I think there are npm libraries that handle this, I would need to look into it.

gambon2010 replied:
'cli-progress' should be a cleaner way to display a progress bar. Looking into implementing it

gambon2010 replied:
completion display is now handled by the  'cli-progress' bar

gambon2010 replied:
'cli-progress' was way too much of a hassle, simplified the printing logic using '\r'

taahamahdi commented on src/generate_json/generate_json.ts:
```diff
@@ -0,0 +1,927 @@
+            (endTime - translationStats.batchStartTime) / 1000,
+        );
+
+        removeLine();
```
Same for these `removeLine`s -- is there a cleaner way to do this?

gambon2010 replied:
'cli-progress' should be a cleaner way to display a progress bar. Looking into implementing it

gambon2010 replied:
completion display is now handled by the  'cli-progress' bar

gambon2010 replied:
'cli-progress' was way too much of a hassle, simplified the printing logic using '\r'

taahamahdi commented on src/generate_csv/generate_csv.ts:
```diff
@@ -1,22 +1,113 @@
+        // eslint-disable-next-line no-await-in-loop
+        const generatedTranslation = await generateTranslation({
+            chats,
+            ensureChangedTranslation: options.ensureChangedTranslation ?? false,
```
Do we need the `??`s if we set defaults?

gambon2010 replied:
`Type 'string | undefined' is not assignable to type 'string'.` typescipt doesn't like the input type.
Maybe we could have an 'input' object in the translate.ts file with optional parameters, and a second object with no optional parameters used in the generate.ts file

gambon2010 replied:
added 'as type' instead

taahamahdi commented on src/generate_json/generate_json.ts:
```diff
@@ -0,0 +1,927 @@
+            {
+                chats,
+                ensureChangedTranslation:
+                    options.ensureChangedTranslation ?? false,
```
Do we need the `??`s if we set defaults?

gambon2010 replied:
added 'as type' instead

taahamahdi commented on src/generate_json/generate_json.ts:
```diff
@@ -0,0 +1,927 @@
+            {
+                chats,
+                ensureChangedTranslation:
+                    options.ensureChangedTranslation ?? false,
```
Do we need the `??`s if we set defaults?

gambon2010 replied:
added 'as type' instead

taahamahdi commented on src/generate_json/generate_json.ts:
```diff
@@ -0,0 +1,927 @@
+    return true;
+}
+
+function getMissingVariables(arr1: string[], arr2: string[]): string[] {
```
Nit: Better naming for arguments, and we could move this into utils

gambon2010 replied:
done

gambon2010 commented on src/cli.ts:
```diff
@@ -45,6 +49,27 @@ const processModelArgs = (options: any): ModelArgs => {
+            }
+
+            if (!options.batchSize) {
+                batchSize = 16;
```
The reason why I added different batch sizes for ollama is that the rate of failed translations goes up by a lot with with a batch size of 32. 
Batch size 16: Translations attempts 900 for 892 lines - Translation time 902 seconds
Batch size 32: Translations attempts 1032 for 892 lines - Translation time 1146 seconds

140 lines fail to translate compared to 8.
Chatgpt doesn't have this issue so I left it at 32 lines per batch.

gambon2010 commented on src/constants.ts:
```diff
@@ -14,18 +15,22 @@ export const DEFAULT_MODEL = {
         "The host and port number serving Ollama. 11434 is the default port number.",
     OverridePromptFile: `Use the prompts from the given JSON file, containing keys for at least one of ${OVERRIDE_PROMPT_KEYS.join(", ")}`,
+    PromptMode:
+        "Chose the prompting mode, between 'csv' mode (better performance but will only work with advanced models like GPT-40) or 'json' mode (compatible with less advanced models, like llama3.1:8b, but translations take 50% longer on average)",
```
fixed in next commit

gambon2010 commented on src/print_styles.ts:
```diff
@@ -0,0 +1,37 @@
+// styles.js
```
fixed in next commit

gambon2010 commented on src/print_styles.ts:
fixed in next comit

gambon2010 commented on src/generate_json/generate_json.ts:
```diff
@@ -0,0 +1,927 @@
+    chats: Chats,
+    translationStats: TranslationStats,
+): Promise<{ [key: string]: string }> {
+    const tikToken = new Tiktoken(
```
Officially tiktoken only supports openai models, but token size is roughly the same for other models. In my current implementation, limiting the size of the prompt depending on available tokens for:
-Smaller models might not have enough tokens left to translate all lines sent.
-The failure message can be pretty long sometimes, especially if multiple lines failed.

And being able to control the size of a prompt will be useful with future Improvements I'm planning, most notably a 'think' tag for the translation. Sort of how deepseek works, if you let the model 'think' before performing a task, the response is generally better, at the cost of execution time/tokens. So having the ability to control token usage more accurately will help avoid sending prompt that are too large, that don;t leave enough tokens for the reply.

Again, this isn't an issue with gpt-40, but does cause many issues with small models.

I'm also going to see if I can find tokenizers for other models

gambon2010 commented on src/cli.ts:
```diff
@@ -45,6 +49,27 @@ const processModelArgs = (options: any): ModelArgs => {
+            } else {
+                promptMode = options.promptMode;
+                if (promptMode === PromptMode.CSV) {
+                    console.warn("WARNING: Json mode recommended for Gemini");
```
fixed in next commit

gambon2010 commented on src/generate_csv/generate_csv.ts:
fixed in next commit

gambon2010 commented on src/chat_interface/anthropic.ts:
```diff
@@ -67,7 +72,12 @@ export default class Anthropic extends ChatInterface {
             return responseText;
         } catch (err) {
-            console.error(err);
+            console.error(
```
Didn't see the second part about the 'subsequent PR'. I updated all the console.log/infos/warn/error to use a function to print with styles.
If you still want to split the PR, I can revert the changes and create an other pull request only with those changes

gambon2010 commented on src/translate.ts:
```diff
@@ -63,30 +58,39 @@ export async function translate(options: TranslateOptions): Promise<Object> {
-    };
-
+function replaceNewlinesWithPlaceholder(
+    options: TranslateOptions,
```
fixed

gambon2010 commented on package.json:
```diff
@@ -8,10 +8,14 @@
     "ollama": "^0.5.12",
     "openai": "^4.27.1",
-    "typescript": "^5.1.6"
+    "tiktoken": "^1.0.20",
```
explained on an other comment why I added it:

`Officially tiktoken only supports openai models, but token size is roughly the same for other models. In my current implementation, I limit the size of a prompt by total tokens because:
-Smaller models might not have enough tokens left to translate all lines sent. (If I'm not mistaken, llama3.1:8b has 4k tokens for the prompt + the response, whereas chatgpt-40 has a 128k token limit for the prompt and can then reply with a maximum of 4k tokens)
-The failure message can be pretty long sometimes, especially if multiple lines failed, which uses a decent amount of tokens.

And being able to control the size of a prompt will be useful with future Improvements I'm planning, most notably a 'think' tag for the translation. Sort of how deepseek works, if you let the model 'think' before performing a task, the response is generally better, at the cost of execution time/tokens. So having the ability to control token usage more accurately will help avoid sending prompts that are too large, that don't leave enough tokens for the response.

Again, this isn't an issue with gpt-40, but does cause many issues with small models.

I'm also going to see if I can find tokenizers for other models`

gambon2010 replied:
Being able to get a token estimation will be somewhat important with the improvements I'm planning on making in JSON mode.
Most notably adding a "think" mode (I detailed it a bit more in one of the other comments). Being able to estimate the tokens used will be crucial.

gambon2010 commented on src/interfaces/options.ts:
```diff
@@ -14,7 +15,9 @@ export default interface Options {
     verbose?: boolean;
     ensureChangedTranslation?: boolean;
     batchSize?: number;
+    batchMaxTokens?: number;
```
By default, an ollama request has a max size of 2048 tokens, even if the model supports more. I think the app should handle this. But I do agree that my current implementation should be improved

gambon2010 commented on src/cli.ts:
```diff
@@ -45,6 +49,27 @@ const processModelArgs = (options: any): ModelArgs => {
+                batchSize = options.batchSize;
+            }
+
+            if (!options.batchMaxTokens) {
```
added error handling after the switch case

gambon2010 commented on src/generate_csv/generate_csv.ts:
```diff
@@ -1,22 +1,113 @@
+
+    for (let i = 0; i < Object.keys(flatInput).length; i += batchSize) {
+        if (i > 0 && options.verbose) {
+            console.info(
```
going to see if I can also display progression with 'cli-progress'

gambon2010 replied:
completion display is now handled by the  'cli-progress' bar

taahamahdi commented on src/translate.ts:
```diff
@@ -122,87 +130,141 @@ export async function translate(options: TranslateOptions): Promise<Object> {
+    switch (options.promptMode) {
+        case PromptMode.JSON:
+            if (options.verbose) {
+                printInfo("Transaltion prompting mode: JSON\n");
```
```suggestion
                printInfo("Translation prompting mode: JSON\n");
```

taahamahdi replied:
Nit: can pull this out of the switch statement, and do `Translation prompting mode: ${options.promptMode}`. Also, is the newline needed?

gambon2010 replied:
Sure

taahamahdi reviewed (changes requested):
Some things I noticed when testing:
* `--verbose` isn't printing outputs in JSON mode
* Using `--prompt-mode json` with `--skip-styling-verification` leads to the error message `throw new Error("'--skip-styling-verification' is not used in CSV mode");`
* Potentially a newline after the ETA: I see the following often:
```
Step 1/2 - Translating | Completed 3% | ETA: 137sRateLimiter | Waiting 2025ms before next API call
```
```
Step 1/2 - Translating | Completed 100% | ETA: 0sTranslation execution time: 2 seconds
```

taahamahdi commented:
This PR has gotten quite long. I don't mind merging it in and addressing the above comments in subsequent smaller PRs, and holding off on a new version in the meantime. What do you think?

gambon2010 commented:
> This PR has gotten quite long. I don't mind merging it in and addressing the above comments in subsequent smaller PRs, and holding off on a new version in the meantime. What do you think?

I agree, as soon as this PR passes I'll fix the rest.

For --verbose, I didn't add print statement for the outputs. In the next pull request I think we should print the ETA even when verbose is off, and only use verbose for printing outputs.

Verification is handled in a single step with JSON mode, so i added an error message when we try to use --skip-styling as it is not used.

I'll fix the new line issue.
